src/Symfony/Component/Cache/Adapter/AbstractAdapter.php

```diff
@@ -127,14 +131,19 @@ public static function createSystemCache($namespace, $defaultLifetime, $version,
 
     public static function createConnection($dsn, array $options = array())
     {
-        if (!is_string($dsn)) {
-            throw new InvalidArgumentException(sprintf('The %s() method expect argument #1 to be string, %s given.', __METHOD__, gettype($dsn)));
-        }
-        if (0 === strpos($dsn, 'redis://')) {
-            return RedisAdapter::createConnection($dsn, $options);
+        @trigger_error(sprintf('The %s() method is deprecated since version 3.4 and will be removed in 4.0. Use the DsnFactory::createConnection() method from Dsn component instead.', __METHOD__), E_USER_DEPRECATED);
+
+        try {
+            $type = ConnectionFactory::getType($dsn);
+        } catch (DsnInvalidArgumentException $e) {
+            throw new InvalidArgumentException($e->getMessage(), 0, $e);
         }
-        if (0 === strpos($dsn, 'memcached://')) {
-            return MemcachedAdapter::createConnection($dsn, $options);
+
+        switch ($type) {
```

Comment:
Should this `switch` logic be moved to the Dsn component itself to avoid duplicating it in every component using it? Something like this:

```php
// Before
switch ($type) {
    case ConnectionFactory::TYPE_MEMCACHED:
        return MemcachedConnectionFactory::createConnection($dsn, $options);
    case ConnectionFactory::TYPE_REDIS:
        return RedisConnectionFactory::createConnection($dsn, $options);
}

// After
return ConnectionFactory($type, $dsn, $options);
```

Reply:
such methods already exists in the Dsn component => `ConnectionFactory::create($dsn, $options)`.

But this methods may return anything (\Pdo, \Ldap, etc...). By using directly the Dsn factory, we are changing the behavior of this method (which previously throws an exception when the DSN wasn't neither `redis://`, neither `memcached://`)

Reply:
I still don't understand this ... but if Nicolas and you say it's correct, then I'm sure it's correct 😄

Reply:
Given the following userland app code
```
try {
   $c = AbstractAdapter::createConnection('mysql://localhost');
  echo 'got a '.get_class($c);
} catch (InvalidArgumentException $e) {
  echo 'expected faillure';
}
```

using the switch => `expected faillure`
using Dsn\ConnectionFactory::create => `got a \Pdo`

It sounds like a BC Break, but IMO i'm also fine to use whats you suggest and saying "we are not responsible of the output of this method if you enter weird things"